lightning/src/ln/onchaintx.rs

```diff
@@ -612,17 +612,15 @@ impl<ChanSigner: ChannelKeys> OnchainTxHandler<ChanSigner> {
 								chan_utils::get_revokeable_redeemscript(&chan_keys.revocation_key, *on_counterparty_tx_csv, &chan_keys.broadcaster_delayed_payment_key)
 							};
 
-							if let Ok(sig) = self.key_storage.sign_justice_transaction(&bumped_tx, i, *amount, &per_commitment_key, htlc, &self.secp_ctx) {
-								bumped_tx.input[i].witness.push(sig.serialize_der().to_vec());
-								bumped_tx.input[i].witness[0].push(SigHashType::All as u8);
-								if htlc.is_some() {
-									bumped_tx.input[i].witness.push(chan_keys.revocation_key.clone().serialize().to_vec());
-								} else {
-									bumped_tx.input[i].witness.push(vec!(1));
-								}
-								bumped_tx.input[i].witness.push(witness_script.clone().into_bytes());
-							} else { return None; }
-							//TODO: panic ?
+							let sig = self.key_storage.sign_justice_transaction(&bumped_tx, i, *amount, &per_commitment_key, htlc, &self.secp_ctx).expect("sign justice tx");
```

Comment:
I do have a doubt on this change, I'm not sure an external signer failure should bring down your Lightning node. You might a partial disk failure, e.g corrupted key data for channel A, but still your signer might still succeed to produce signatures for channel B. 

Or this a kind of scenario which should be managed behind the key interface and a dummy sig should be returned ?

Reply:
Here are my thoughts on this.

I believe that a highly available installation should have replication at these layers:

- signer - there should be a quorum of signers that have to agree for channel state advancement
- chanmon - there should be at least one chanmon replica reacting to on-chain events
- network handler (everything other than signer/chanmon) - probably want this in a active+hot-standby configuration, but maybe you can even do active+active or split by channel

I think in such a configuration, the signer layer should have a "TEMPFAIL" result, which should defer the operation and retry later.  For example, there may not be a quorum for one channel, but another channel can go ahead.  I think this requires some additional code to implement the deferral, so should be its own effort.  However, the existing errors this function produced are not in that category, they indicated a bug or corruption.

I think a faulty disk should probably result in the replica brought down and operator intervention, since it's hard to automatically figure out if the replica is still sane (maybe we are seeing old data, maybe our RAM is bad).  In a replicated environment, there should be enough other replicas to continue.

To summarize, I think the existing errors should be fatal, but in the future we should create a new kind of temporary error in all signer API calls.  This kind of error should be generated by the signer "client-side" if it fails to arrange for a quorum and should cause the operation to be deferred.

Reply:
> You might a partial disk failure, e.g corrupted key data for channel A, but still your signer might still succeed to produce signatures for channel B.

I'd strongly suggest that this is a *great* reason to bring down a lightning node - your hardware is corrupting data, for all we know the data may be corrupted in a way that results in us revealing a private key.

Reply:
I like the idea of temporary errors generated by the "client-side" signer, I agree that's a good direction for the future.

> your hardware is corrupting data, for all we know the data may be corrupted in a way that results in us revealing a private key.

It's hard to say, on one hand, if you have only one replica and you're crashing your only monitoring instance you might lose all your channels instead of burning only one. And the other hand, your hardware failure might be so gross that you reveal private key of L1 funds if your signer is cross-layer. 

I don't think there is _a_ good answer and it's likely function of your configuration, funds at stake, confidence in your hardware, ... One more trade-off we will have to document somewhere I guess.

Reply:
Sure, there isn't a clear answer, but we do know its critical that the user get involved ASAP and figure out what issue is happening. Continuing on is usually a great way to not have the user get involved, and I'd argue that the only real way we have of doing that is to panic.